test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/ScalarReplacementWithGCBarrierTests.java

```diff
@@ -0,0 +1,118 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.irTests.scalarReplacement;
+
+import compiler.lib.ir_framework.*;
+
+/*
+ * @test
+ * @bug 8333334
+ * @summary Tests that dead barrier control flows do not affect the scalar replacement.
+ * @library /test/lib /
+ * @requires vm.compiler2.enabled
+ * @requires vm.gc.G1
+ * @run driver compiler.c2.irTests.scalarReplacement.ScalarReplacementWithGCBarrierTests
+ */
+public class ScalarReplacementWithGCBarrierTests {
+    static class List {
+        public Node head;
+
+        public void push(int value) {
+            Node n = new Node();
+            n.value = value;
+            n.next = head;
+            head = n;
+        }
+
+        @ForceInline
+        public Iter iter() {
+            Iter iter = new Iter();
+            iter.list = this;
+            iter.n = head;
+            iter.sum = 0;
+            return iter;
+        }
+    }
+
+    static class Node {
+        public int value;
+        public Node next;
+    }
+
+    static class Iter {
+        public List list;
+        public Node n;
+        public Integer sum;
+
+        @ForceInline
+        public boolean next() {
+            int lastSum = sum;
+            while (sum - lastSum < 1000) {
+                while (n != null && n.value < 30) n = n.next;
+                if (n == null) return false;
+                sum += n.value;
+                n = n.next;
+            }
+            return true;
+        }
+    }
+
+    private static final int SIZE = 1000;
+
+    public static void main(String[] args) {
+        // Must use G1 GC to ensure there is a pre-barrier
+        // before the first field write.
+        TestFramework.runWithFlags("-XX:+UseG1GC");
+    }
+
+    @Run(test = "testScalarReplacementWithGCBarrier")
+    private void runner() {
+        List list = new List();
+        for (int i = 0; i < SIZE; i++) {
+            list.push(i);
+        }
+        testScalarReplacementWithGCBarrier(list);
+    }
+
+    // Allocation of `Iter iter` should be eliminated by scalar replacement, and
+    // the allocation of `Integer sum` can not be eliminated, so there should be
+    // 1 allocation after allocations and locks elimination.
+    //
+    // Before the patch of JDK-8333334, both allocations of `Iter` and `Integer`
+    // could not be eliminated.
+    @Test
+    @IR(phase = { CompilePhase.AFTER_PARSING }, counts = { IRNode.ALLOC, "1" })
+    @IR(phase = { CompilePhase.INCREMENTAL_BOXING_INLINE }, counts = { IRNode.ALLOC, "2" })
+    @IR(applyIf = { "UseG1GC", "true" }, phase = { CompilePhase.ITER_GVN_AFTER_ELIMINATION }, counts = { IRNode.ALLOC, "1" })
```

Comment:
Your test checks for the number of allocations to return to one. However in your description and comments you talk about load nodes that aren't folded. What about adding another IR test to check the number of loads for completeness?

Reply:
I think it's hard to add an intuitive and meaningful IR test for load nodes in this test case.

To construct a case that containing both eliminable GC pre-barriers and allocations, and have them interfere with each other, I had to put several object field reads/writes into nested loops. This makes a large number of load nodes appearing in this test case after some optimization phases. Perhaps after several more optimizations, most of them will disappear and new loads will appear again. So it's difficult to tell which load we really care about without looking at the Ideal graph dump via IGV.

Reply:
Right, I think the test is fine as is.